Source/WebGPU/WGSLUnitTests/Parser/TypeDeclTests.mm

```diff
@@ -0,0 +1,168 @@
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#import "LiteralExpressions.h"
+#import "ParserPrivate.h"
+#import <XCTest/XCTest.h>
+
+static Expected<UniqueRef<WGSL::AST::TypeDecl>, WGSL::Error> parseLCharTypeDecl(const String& input)
+{
+    WGSL::Lexer<LChar> lexer(input);
+    WGSL::Parser parser(lexer);
+
+    return parser.parseTypeDecl();
+}
+
+static void testNamedType(const String& type)
+{
+    auto declResult = parseLCharTypeDecl(type);
+    XCTAssert(declResult);
+    const auto& decl = (*declResult).get();
+
+    XCTAssert(decl.isNamed());
+    const auto& namedDecl = downcast<WGSL::AST::NamedType>(decl);
+
+    XCTAssert(namedDecl.name() == type);
+}
+
+@interface TypeDeclTests : XCTestCase
+
+@end
+
+@implementation TypeDeclTests
+
+- (void)setUp {
+    [super setUp];
+    self.continueAfterFailure = NO;
+}
+
+- (void)testBool {
+    testNamedType("bool"_s);
+}
+
+- (void)testI32 {
+    testNamedType("i32"_s);
+}
+
+- (void)testU32 {
+    testNamedType("u32"_s);
+}
+
+- (void)testF32 {
+    testNamedType("f32"_s);
+}
+
+- (void)testIdentifier {
+    testNamedType("some_type_alias"_s);
+}
+
+- (void)testVec4 {
+    auto declResult = parseLCharTypeDecl("vec4<f32>"_s);
+    XCTAssert(declResult);
+    auto decl = WTFMove(*declResult);
+
+    XCTAssert(decl->isVec());
+    const auto& vecDecl = downcast<WGSL::AST::VecType>(decl.get());
+
+    XCTAssert(vecDecl.size() == 4);
+
+    const auto& elementType = vecDecl.elementType();
+    XCTAssert(elementType.isNamed());
+    const auto& namedElementType = downcast<WGSL::AST::NamedType>(elementType);
+
+    XCTAssert(namedElementType.name() == "f32"_s);
+}
+
+- (void)testVecWithoutElementType {
+    auto declResult = parseLCharTypeDecl("vec4"_s);
+    XCTAssert(!declResult);
+    // FIXME: check error is correct.
```

Comment:
How much effort is required to add a check for the expected error?

Reply:
Not much - this is just an initial patch so someone can check the parsing logic. I'll update it with the check later.

Reply:
Sounds good.